pkgs/build-support/setup-hooks/make-binary-wrapper.sh

```diff
@@ -0,0 +1,260 @@
+                uses_assert=1
+                n=$((n + 1))
+                [ $n -ge "$length" ] && main="$main    #error makeCWrapper: $p takes 1 argument"$'\n'
+            ;;
+            --argv0)
+                argv0=$(escapeStringLiteral "${params[n + 1]}")
+                n=$((n + 1))
+                [ $n -ge "$length" ] && main="$main    #error makeCWrapper: $p takes 1 argument"$'\n'
+            ;;
+            *) # Using an error macro, we will make sure the compiler gives an understandable error message
+                main="$main    #error makeCWrapper: Uknown argument ${p}"$'\n'
+            ;;
+        esac
+    done
+    # shellcheck disable=SC2086
+    [ -z "$flagsBefore" ] || main="$main"${main:+$'\n'}$(addFlags $flagsBefore)$'\n'$'\n'
+    main="$main    argv[0] = \"${argv0:-${executable}}\";"$'\n'
+    main="$main    return execv(\"${executable}\", argv);"$'\n'
+
+    printf '%s\n' "#include <unistd.h>"
+    printf '%s\n' "#include <stdlib.h>"
+    [ -z "$uses_concat3" ] || printf '%s\n' "#include <string.h>"
+    [ -z "$uses_assert" ]  || printf '%s\n' "#include <assert.h>"
+    [ -z "$uses_concat3" ] || printf '\n%s\n' "$(concat3Fn)"
+    [ -z "$uses_prefix" ]  || printf '\n%s\n' "$(setEnvPrefixFn)"
+    [ -z "$uses_suffix" ]  || printf '\n%s\n' "$(setEnvSuffixFn)"
+    printf '\n%s' "int main(int argc, char **argv) {"
+    printf '\n%s' "$main"
+    printf '%s\n' "}"
+}
+
+addFlags() {
+    local result n flag flags var
+    var="argv_tmp"
+    flags=("$@")
+    for ((n = 0; n < ${#flags[*]}; n += 1)); do
+        flag=$(escapeStringLiteral "${flags[$n]}")
+        result="$result    ${var}[$((n+1))] = \"$flag\";"$'\n'
+    done
+    printf '    %s\n' "char **$var = malloc(sizeof(*$var) * ($((n+1)) + argc));"
+    printf '    %s\n' "assert($var != NULL);"
+    printf '    %s\n' "${var}[0] = argv[0];"
+    printf '%s' "$result"
+    printf '    %s\n' "for (int i = 1; i < argc; ++i) {"
+    printf '    %s\n' "    ${var}[$n + i] = argv[i];"
+    printf '    %s\n' "}"
+    printf '    %s\n' "${var}[$n + argc] = NULL;"
+    printf '    %s\n' "argv = $var;"
+}
+
+# prefix ENV SEP VAL
+setEnvPrefix() {
+    local env sep val
+    env=$(escapeStringLiteral "$1")
+    sep=$(escapeStringLiteral "$2")
+    val=$(escapeStringLiteral "$3")
+    printf '%s' "set_env_prefix(\"$env\", \"$sep\", \"$val\");"
+}
+
+# suffix ENV SEP VAL
+setEnvSuffix() {
+    local env sep val
+    env=$(escapeStringLiteral "$1")
+    sep=$(escapeStringLiteral "$2")
+    val=$(escapeStringLiteral "$3")
+    printf '%s' "set_env_suffix(\"$env\", \"$sep\", \"$val\");"
+}
+
+# setEnv KEY VALUE
+setEnv() {
+    local key value
+    key=$(escapeStringLiteral "$1")
+    value=$(escapeStringLiteral "$2")
+    printf '%s' "putenv(\"$key=$value\");"
+}
+
+# setDefaultEnv KEY VALUE
+setDefaultEnv() {
+    local key value
+    key=$(escapeStringLiteral "$1")
+    value=$(escapeStringLiteral "$2")
+    printf '%s' "setenv(\"$key\", \"$value\", 0);"
```

Comment:
I'm guessing we want to crash the program if setenv fails, so how about something like this?
```sh
printf '%s' "assert(setenv(\"$key\", \"$value\", 0) == 0);"
```

As far as I can tell there are 2 reasons setenv might fail: insufficient memory and variable name containing `=`, which we can catch during code generation instead of during runtime.

Reply:
that sounds reasonable. though if we're adding enough asserts it might also be nice to have a version of it that also prints errno, basically `#define assert_success(e) do { if ((e) < 0) { perror(#e); exit(1); } while (0)`

Reply:
I `exit(1)` *equivalent enough* with an `assert()` return value?

I'm reading it's likely to be 134 (128 + SIGABRT), tbf I haven't tested.

Is there value in ensuring it exits in a similar manner?

Reply:
it's more likely to be equivalent to `raise(SIGABRT)`, which would indeed be better than using exit!